security/pkg/k8s/chiron/utils.go

```diff
@@ -313,15 +313,25 @@ func readSignedCertificate(client clientset.Interface, csrName string,
 		if ok := roots.AppendCertsFromPEM(caCert); !ok {
 			return nil, nil, fmt.Errorf("failed to append CA certificate")
 		}
-		_, err = certParsed.Verify(x509.VerifyOptions{
-			Roots: roots,
+
+		intermediates := x509.NewCertPool()
+		if len(certsParsed) > 1 {
+			for _, cert := range certsParsed[1:] {
+				intermediates.AddCert(cert)
+			}
+		}
+
+		_, err = certsParsed[0].Verify(x509.VerifyOptions{
+			Roots:         roots,
+			Intermediates: intermediates,
 		})
 		if err != nil {
 			return nil, nil, fmt.Errorf("failed to verify the certificate chain: %v", err)
 		}
-		certChain = append(certChain, caCert...)
+		certPEM = append(certPEM, caCert...)
```

Comment:
Question: Why are we appending the CA certificate to the chain here? Is this not shared out of band with cert gRPC clients?

Reply:
Not really. Today, the only way that the mesh infers trustAnchors for in-mesh mTLS is by looking at the root of the signed workload cert chain. Since the initial signer was the default "legacy-unknown" signer who did not return the entire  cert-chain (including the CA-cert), this code was relevant. 

For most custom signers, we don't need to append the ca cert since it will be a part of the cert-chain. We could have a env configuration to control this

Reply:
I see. It is surprising to me (and concerning?) that clients are intended to blindly trust whatever CA is returned here. Would this not be a vector for abuse?

I assumed all CA trust was done via the config map propagation by istiod or other means.

Very well could be well off here.

Reply:
If a signer was able to hijack CSRs and istiod did not have `appendCaCert` enabled, can a malicious actor start making clients trust whatever they want?

Could this also be true even if `appendCACert` is enabled?

Reply:
It is a valid concern, although not sure how it necessarily opens up a larger attack surface area. Also, our CA functionality was limited until now, since we mainly relied on citadel in OSS. With the RA functionality, this question becomes more relevant. 

Note that we also configure additional trustAnchors through the MeshConfig config-map, though this is newer functionality. Perhaps our default behavior should be to check if the root of the returned cert-chain from the CA matches with one of these trustAnchors...

Reply:
going point reg the malicious signer. Reinforces my idea that when using Istiod as RA, we mandate that the signer should affix their CA cert to the signed certChain and we compare this against our own configured TrustAnchors (in meshConfig or mounted to Istiod)

Reply:
> Perhaps our default behavior should be to check if the root of the returned cert-chain from the CA matches with one of these trustAnchors...

This makes sense to me :+1:

Reply:
> going point reg the malicious signer. Reinforces my idea that when using Istiod as RA, we mandate that the signer should affix their CA cert to the signed certChain and we compare this against our own configured TrustAnchors (in meshConfig or mounted to Istiod) to confirm the identity of the CA

This doesn't require the signer to send out their CA though right? The RA should be able to correctly verify a cert chain without the CA since it already has it as an entry in the TrustAnchors.. Just the same as normal TLS.

Reply:
Coming from the cert-manager side, we are apprehensive about including the CA in the chain response as it basically allows people to shoot themselves in the foot by allowing them to incorrectly blindly trust that CA.

Whilst sending the CA isn't bad in isolation, not sending it forces users to do the right thing.

Reply:
> It is a valid concern, although not sure how it necessarily opens up a larger attack surface area. Also, our CA functionality was limited until now, since we mainly relied on istiod. With the RA functionality, this question becomes more relevant.
> 
> Note that we also configure additional trustAnchors through the MeshConfig config-map, though this is newer functionality. Perhaps our default behavior should be to check if the root of the returned cert-chain from the CA matches with one of these trustAnchors...

Does this mean admin needs to upload all of their root certs in meshconfig with mapping for the signers? seems our current mesh config API does not support this, we may need to change it and think about how to distribute them efficiently.